File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    w, h = len(park[0]), len(park)
    dp = [[0 if park[0][x] != "-1" else 1 for x in range(w)]] + [[0 for _ in range(w)] for _ in range(h-1)]
    big = max(dp[0])
    for r in range(1, h):
        for c in range(w):
            if not c and park[r][0] == "-1":
                dp[r][0] = 1
            if c and park[r][c] == "-1":
                dp[r][c] = min(dp[r-1][c], dp[r][c-1], dp[r-1][c-1]) +1
            big = max(big, dp[r][c])
    answer = -1
    for l in mats:
        if l <= big:
            answer = max(answer, l)
    return answer

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
import java.math.*;

class Solution {
    public int solution(int[] mats, String[][] park) {
        int answer = -1;
        int h = park.length; int w = park[0].length;
        int[][] dp = new int[h][w];
        int big = 0;
        for (int r = 0; r < h; r++){
            for (int c = 0; c < w; c++){
                if (park[r][c].equals("-1")){
                    if (r == 0) dp[0][c] = 1;
                    else if (c == 0) dp[r][0] = 1;
                    else dp[r][c] = Math.min(Math.min(dp[r-1][c], dp[r][c-1]), dp[r-1][c-1]) +1;
                    big = Math.max(big, dp[r][c]);
                }
            }
        }
        
        for (int l:mats){
            if (l <= big) answer = Math.max(answer, l);
        }
        
        
        return answer;
    }
}